src/main/scala/scalajsGames/games/AstroLander.scala

```diff
@@ -78,77 +79,94 @@ case class AstroLander(bounds: Point, resetGame: () => Unit) extends Game{
     ctx.moveTo(shipPoints.last.x, shipPoints.last.y)
     shipPoints.map(p => ctx.lineTo(p.x, p.y))
     ctx.fill()
+
     def drawFlame(p: Point, angle: Double) = {
       val offset = math.Pi * 1.25
+
       def diamond(a: Int, b: Int, c: Int, w: Int) = {
         val width = w * math.Pi / 180
 
         ctx.strokePath(
           p + Point(a, a).rotate(angle - offset),
           p + Point(b, b).rotate(angle - offset + width),
           p + Point(c, c).rotate(angle - offset),
-          p + Point(b, b).rotate(angle - offset  - width),
+          p + Point(b, b).rotate(angle - offset - width),
           p + Point(a, a).rotate(angle - offset)
         )
       }
+
       diamond(5, 15, 25, 15)
       diamond(10, 15, 20, 10)
     }
 
     ctx.strokeStyle = Color.Red
-    if (fuel > 0){
+    if (fuel > 0) {
       if (lastKeys(37)) drawFlame(craftPos, theta + math.Pi / 4)
       if (lastKeys(39)) drawFlame(craftPos, theta - math.Pi / 4)
       if (lastKeys(40)) drawFlame(craftPos, theta)
     }
 
   }
+
   var lastKeys: Set[Int] = Set()
+
   def update(keys: Set[Int]) = {
     lastKeys = keys
-    if (fuel > 0){
-      if (keys(37)) craftVel += Point(0.5, 0).rotate(theta + math.Pi / 4)
-      if (keys(39)) craftVel += Point(0.5, 0).rotate(theta - math.Pi / 4)
-      if (keys(40)) craftVel += Point(0.5, 0).rotate(theta)
+    if (fuel > 0) {
+      // I changed the first parameter of Point of 0.5 to 0.1, thats makes the ship slowest
+      if (keys(37)) craftVel += Point(0.1, 0).rotate(theta + math.Pi / 4)
+      if (keys(39)) craftVel += Point(0.1, 0).rotate(theta - math.Pi / 4)
+      if (keys(40)) craftVel += Point(0.4, 0).rotate(theta)
       fuel -= Seq(keys(37), keys(39), keys(40)).count(x => x)
     }
 
     craftVel += Point(0, 0.2)
     craftPos += craftVel
 
+    val shipLost = craftPos.x < 0 || craftPos.x >= bounds.x || craftPos.y <= 1 || craftPos.y >= bounds.y
+    val hit = points.flatMap { p =>
+      val prevIndex = points.lastIndexWhere(_.x < craftPos.x)
 
 
-    val hit = points.flatMap{ p =>
-      val prevIndex = points.lastIndexWhere(_.x < craftPos.x)
-      if (prevIndex == -1 || prevIndex == 21) None
-      else{
-        val prev = points(prevIndex)
+      if (prevIndex == 21) None
+      else {
+        val prev = points(if prevIndex == -1 then prevIndex + 1 else prevIndex)
         val next = points(prevIndex + 1)
         val height = (craftPos.x - prev.x) / (next.x - prev.x) * (next.y - prev.y) + prev.y
-        if (height > craftPos.y) None
-        else Some{
+        if (height > craftPos.y && !shipLost) None
+        else Some {
           val groundGradient = math.abs((next.y - prev.y) / (next.x - prev.x))
-          val  landingSkew = math.abs(craftVel.x / craftVel.y)
+          val landingSkew = math.abs(craftVel.x / craftVel.y)
+
 
-          if (groundGradient > 0.1) Failure("landing area too steep")
+          if (shipLost) ExtraFailure("You have lost in the space")
```

Comment:
Please don't do this, just prevent the ship from getting lost by keeping it at the boundary

Reply:
Done.